File: data_sources/sky_server.rb
require 'httparty'
require 'nokogiri'

class SkyServer
  include HTTParty
  base_uri 'skyserver.sdss3.org'

  @@photo_fields = %w(objid ra dec b l u g r i z petroR90_u petroR90_g petroR90_i petroR90_r petroR90_z)
  @@spec_fields = %w(plate mjd fiberID z specObjID)

  def self.fetch query
    get '/dr9/en/tools/search/x_sql.asp', :query => { :format => 'xml',
                                          :cmd => query }
  end

  def self.by_ra_dec options
    query = %Q(SELECT TOP #{options[:limit]}
              #{@@photo_fields.map{|field| "g.#{field}"}.join(',')}
              #{", #{@@spec_fields.map{|field| "s.#{field}"}.join(',')} as redshift" if options[:spec]}
              FROM Galaxy as g, dbo.fGetNearbyObjEq(#{options[:ra]},#{options[:dec]},#{options[:radius]}) as n
              #{"JOIN SpecObj as s ON s.bestobjid = n.objid" if options[:spec]}
              WHERE n.objid=g.objid).gsub("\n", "")
    self.query query
  end

  def self.by_ugriz options
    where = [:u, :i, :r, :g, :z].select{|band| !options[band].nil?}
      .map{ |band| "g.#{band.to_s} BETWEEN #{options[band].to_f - options[:tolerance]} AND #{options[band].to_f + options[:tolerance]}"}
      .join(" AND ")
    query = %Q(SELECT TOP #{options[:limit]} #{@@photo_fields.map{|field| "g.#{field}"}.join(',')}
                #{", #{@@spec_fields.map{|field| "s.#{field}"}.join(',')} as redshift" if options[:spec]}
                FROM  Galaxy as g
                #{"JOIN SpecObj as s ON s.bestobjid = g.objid" if options[:spec]}
                WHERE #{where})
    self.query query
  end

  def self.query query
    begin
      res = fetch query
      format res.parsed_response['root']['Answer']['Row']
    rescue Exception => e
      "Error: #{e.message}"
    end
  end

  def self.format data
    data.each do |subject|
      subject.each do |key, value|
        subject[key] = value.to_f unless (key =='objid') || (@@spec_fields.include? key)
      end
      subject['image'] = image_uri subject['ra'], subject['dec']
      subject['uid'] = subject['objid']
    end
  end

  def self.image_uri ra, dec
    %Q(http://skyservice.pha.jhu.edu/DR9/ImgCutout/getjpeg.aspx?ra=#{ra}&dec=#{dec}&scale=0.15&width=300&height=300&opt=)
  end

end

File: src/spelunker/model/sky_server.clj
(ns spelunker.model.sky-server
  (:require [spelunker.model.mongo :as m]
            [clojure.string :as str]
            [net.cgrand.enlive-html :as e]))

(def collection "sky-server")

;; Functions to Generate Sky Server Compitable SQL

(defn- select-top
  [limit]
  (str "SELECT TOP " limit " g.objid, g.ra, g.dec, g.b, g.l, g.u, g.r, g.g, g.r, g.i, g.z, g.petroR90_u, g.petroR90_g, g.petroR90_r, g.petroR90_i, g.petroR90_z, s.plate, s.mjd, s.fiberID, s.specObjID, s.z as redshift"))

(def from-galaxy "FROM Galaxy as g")

(defn- with-coords
  [ra dec radius]
  (str ", dbo.fGetNearbyObjEq(" ra "," dec "," radius ") as n"))

(defn- join-to
  [prefix]
  (str "JOIN SpecObj as s ON s.bestobjid = " prefix ".objid"))

(def where-area "WHERE g.objid = n.objid")

(defn where-bands
  [params]
  (str "WHERE " (str/join " AND " (map (fn [[k v]] (str "g." k " " v)) params))))

(defmulti convert-params :search-type)

(defmethod convert-params "area" [{:keys [params]}]
  (let [{:keys [ra dec limit radius]} params]
    (str (select-top limit) " " from-galaxy (with-coords ra dec radius) " " (join-to "n") " " where-area)))

(defmethod convert-params "bands" [{:keys [params]}]
  (let [limit (:limit params)
        params (dissoc params :limit)]
    (str (select-top limit) " " from-galaxy " " (join-to "g") " " (where-bands params))))

(defmethod convert-params "sql" [{:keys [params]}]
  (:query params))

;; Implementation of Querying

(defn- row-to-map
  [row]
  (into {} (map (fn [tag] {(:tag tag) (first (:content tag))}) row)))

(defn- image-for-subject
  [subject]
  (let [{:keys [ra dec]} subject
        image-url (str "http://skyservice.pha.jhu.edu/DR9/ImgCutout/getjpeg.aspx?ra="
                       ra "&dec=" dec "&scale=0.15&width=300&height=300&opt=")]
    (assoc subject :image image-url)))

(defn- parse-sky-server
  "Convert SkyServer XML into clojure map."
  [xml]
  (->> (e/select xml [:Row])
       (map :content)
       (map row-to-map)
       (map (fn [subject]
              (let [uid (:objid subject)]
                (-> (dissoc subject :objid)
                    (assoc :uid uid)))))
       (map image-for-subject)))

(defn run-query
  [query] 
  (let [query (merge query {:params {:format "xml" :cmd (convert-params query)}})]
    (m/run-query collection parse-sky-server
                 "http://skyserver.sdss3.org/dr9/en/tools/search/x_sql.asp"
                 query)))

(def query! (partial m/query! collection))
(def get-query (partial m/get-query collection))
(def add-subjects! (partial m/add-subjects! collection))
(def get-subjects (partial m/get-subjects collection))
